[CS] Some SyntacticElement refactorings

hamishknight opened the pull request:
Split off some refactorings from the if expression work. This should be NFC.

xedin committed:
[CSClosure] Introduce `SyntacticElementContext`

Replace `AnyFunctionRef` as "context" for syntactic element with
a custom `SyntacticElementContext` to support type-checking of
other constructs in the future.

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

hamishknight commented:
@swift-ci please test Linux

xedin commented on lib/Sema/BuilderTransform.cpp:
```diff
@@ -364,68 +364,70 @@ class BuilderClosureVisitor
 
-  VarDecl *visitBraceStmt(BraceStmt *braceStmt) {
-    SmallVector<Expr *, 4> expressions;
+  void visitBraceElement(ASTNode node, SmallVectorImpl<Expr *> &expressions) {
```
Let’s not do this as BuilderClosureTransform and all relative logic is going away completely soon - https://github.com/apple/swift/pull/60419

xedin commented on lib/Sema/CSSyntacticElement.cpp:
```diff
@@ -358,16 +358,24 @@ class SyntacticElementConstraintGenerator
 
   ConstraintSystem &cs;
-  AnyFunctionRef context;
+  DeclContext *dc;
```
This is actually moving in opposite direction (it used to be DeclContext originally) but we’d like to support `TapExpr` as a context as well - https://github.com/apple/swift/pull/61746

hamishknight replied:
Makes sense, shall I pull in SyntacticElementContext minus `TapExpr` for this PR?

xedin replied:
Sure, it should be cherry-pickable.

hamishknight replied:
How does this look?

xedin commented on lib/Sema/CSBindings.cpp:
```diff
@@ -1065,6 +1065,14 @@ bool BindingSet::favoredOverDisjunction(Constraint *disjunction) const {
 
+bool BindingSet::favoredOverConjunction(Constraint *conjunction) const {
+  if (CS.shouldAttemptFixes() && isHole()) {
+    if (forClosureResult() || forGenericParameter())
```
Why is this important in diagnostic mode?

hamishknight replied:
No idea, I'm just preserving the original logic :)

xedin replied:
I should be trying to remember then :)

hamishknight committed:
[CS] NFC: Rename CSClosure.cpp -> CSSyntacticElement.cpp

We want to generalize this conjunction logic to
not just be applicable to closures, so rename the
file to match.

hamishknight committed:
[CS] Factor out brace element handling in a couple of cases

Split out brace element handling for a couple of
walkers into a new function. The diff for this is
best viewed without whitespace changes. This
should be NFC.

hamishknight committed:
[CS] Factor out `BindingSet::favoredOverConjunction`

And refactor the way we select the next step for
a ComponentStep. This should be NFC.

hamishknight committed:
[CS] NFC: Remove unused method

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test
